sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs

```diff
@@ -0,0 +1,419 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Azure.Core;
+using Azure.Core.Pipeline;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Azure.Identity
+{
+    internal class ManagedIdentityClient
+    {
+        private const string MsiEndpointEnvironemntVariable = "MSI_ENDPOINT";
+        private const string MsiSecretEnvironemntVariable = "MSI_SECRET";
+        private static Lazy<ManagedIdentityClient> s_sharedClient = new Lazy<ManagedIdentityClient>(() => new ManagedIdentityClient());
+        private static readonly Uri ImdsEndpoint = new Uri("http://169.254.169.254/metadata/identity/oauth2/token");
+        private static SemaphoreSlim s_initLock = new SemaphoreSlim(1, 1);
+        private static MsiType s_msiType;
+        private static Uri s_endpoint;
+        private static string s_secret;
+
+        private const string ImdsApiVersion = "2018-02-01";
+        private const string AppServiceMsiApiVersion = "2017-09-01";
+
+
+        private readonly IdentityClientOptions _options;
+        private readonly HttpPipeline _pipeline;
+
+        public ManagedIdentityClient(IdentityClientOptions options = null)
+        {
+            _options = options ?? new IdentityClientOptions();
+
+            _pipeline = HttpPipelineBuilder.Build(_options, bufferResponse: true);
+        }
+
+        private enum MsiType
+        {
+            Unknown = 0,
+            Imds = 1,
+            AppService = 2,
+            CloudShell = 3,
+            Unavailable = 4
+        }
+
+        public static ManagedIdentityClient SharedClient { get { return s_sharedClient.Value; } }
+
+        public virtual async Task<AccessToken> AuthenticateAsync(string[] scopes, string clientId = null, CancellationToken cancellationToken = default)
+        {
+            MsiType msiType = GetMsiType(cancellationToken);
+
+            // if msi is unavailable or we were unable to determine the type return a default access token
+            if (msiType == MsiType.Unavailable || msiType == MsiType.Unknown)
+            {
+                return default;
+            }
+
+            using (Request request = CreateAuthRequest(msiType, scopes, clientId))
+            {
+                var response = await _pipeline.SendRequestAsync(request, cancellationToken).ConfigureAwait(false);
+
+                if (response.Status == 200 || response.Status == 201)
+                {
+                    var result = await DeserializeAsync(response.ContentStream, cancellationToken).ConfigureAwait(false);
+
+                    return result;
+                }
+
+                throw response.CreateRequestFailedException();
```

Comment:
We should add an exception type so that these failures can be easily distinguished from service client failures, as both would originate from the same user call to a service client method.  I've added issue #6712 to track this.